### PR TITLE
Onchain fulfill and penalty tx

### PR DESCRIPTION
**Pull request by rustyrussell**

### CODE REVIEW

File: tests/utils.py [hunk 1]
```diff
@@ -307,3 +308,9 @@ def db_query(self, query):
         c.close()
         db.close()
         return result
+
+    def has_failed(self):
```
[the remark this answers is not in the document]
Comment (cdecker) [reply tests/utils.py]: Alternatively we could also add a keyword arg to the `node_factory.get_node` to tell it that this node `may_fail`, seems cleaner than to annotate after the fact.

File: onchaind/onchain.c [hunk 2]
```diff
@@ -502,8 +502,55 @@ static void handle_htlc_onchain_fulfill(struct tracked_output *out,
...
+
+	/* BOLT #5:
+	 *
+	 * A node SHOULD resolve its own HTLC transaction output by spending
+	 * it to a convenient address.  A node MUST wait until the
+	 * `OP_CHECKSEQUENCEVERIFY` delay has passed (as specified by the
+	 * other node's `open_channel` `to_self_delay` field) before spending
+	 * the output.
+	 */
+	out = new_tracked_output(outs, htlc_txid, tx_blockheight,
```
Comment (cdecker) [inline onchaind/onchain.c]: We'll probably eventually want to add these to the DB as well so we don't assume that `onchaind` stays up the entire time. So we end up with two launch modes for `onchaind`: a hurry and settle on chain, and a go gather HTLCs mode. I'm guessing the HTLC state should already be sufficient to reconstruct the gathering transactions?

File: tests/test_lightningd.py [hunk 3]
```diff
@@ -472,8 +473,14 @@ def test_onchain_middleman(self):
         route = l1.rpc.getroute(l3.info['id'], 10**8, 1)["route"]
         assert len(route) == 2
 
+        q = queue.Queue()
```
Comment (cdecker) [inline tests/test_lightningd.py]: As mentioned during the last meeting, I'll rewrite this to use futures :wink:

File: lightningd/peer_control.c [hunk 4]
```diff
@@ -1627,6 +1632,51 @@ void peer_last_tx(struct peer *peer, struct bitcoin_tx *tx,
 	peer->last_tx = tal_steal(peer, tx);
 }
 
+static void json_sign_last_tx(struct command *cmd,
```
Comment (cdecker) [inline lightningd/peer_control.c]: Ouch, this is really dangerous. Any chance we can hide this in releases behind a couple of compile guards? See #176

File: onchaind/onchain.c [hunk 5]
```diff
@@ -749,14 +748,44 @@ static void output_spent(struct tracked_output ***outs,
...
+		     tx_type_name(out->resolved->tx_type),
+		     depth);
+
+	/* BOLT #5:
+	 *
+	 * If the HTLC output has *timed out* and not been *resolved*,
+	 * the node MUST *resolve* the output and MUST fail the
+	 * corresponding incoming HTLC (if any) once the resolving
+	 * transaction has reached reasonable depth. */
+	if (out->resolved->tx_type == OUR_HTLC_TIMEOUT_TX
```
Comment (cdecker) [inline onchaind/onchain.c]: Can be unnested

File: channeld/channel_wire.csv [hunk 6]
```diff
@@ -1,11 +1,11 @@
 # Received and sent funding_locked
-channel_normal_operation,1001
+channel_normal_operation,11001
```
Comment (cdecker) [inline channeld/channel_wire.csv]: Why don't we add structure to the messages? We could enumerate the daemons and each daemon accepts `daemon_num | message_num` and fails anything else. That way we wouldn't need to manually make sure that these numbers are unique (and peer messages would simply get `daemon_num = 0`)

File: tests/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rustyrussell) [reply tests/utils.py]: Good idea. We could suppress stderr too in that case

File: onchaind/onchain.c [hunk 7]
```diff
@@ -118,12 +119,44 @@ static void narrow_feerate_range(struct feerate_range *feerate_range,
...
+	if (min > feerate_range.min)
+ 		feerate_range.min = min;
 	status_trace("Feerate now %u to %u",
-		     feerate_range->min, feerate_range->max);
+		     feerate_range.min, feerate_range.max);
+}
+
+/* We vary feerate until signature they offered matches: we're more
+ * likely to be near max. */
+static bool grind_feerate(struct bitcoin_tx *commit_tx,
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply onchaind/onchain.c]: Worst case is where we have what looks like a huge fee, due to dust output trimming.  We have 483*2 htlcs of amount dust_limit-1.  If the dust limit is 10k satoshi, that's 10M iterations.  We can certainly optimize this code, I've opened an issue to ensure I do proper benchmarking #291

File: lightningd/peer_control.c (hunk 4 above)
Comment (rustyrussell) [reply lightningd/peer_control.c]: Yes, definitely!  Will add FIXME here.

File: channeld/channel_wire.csv (hunk 6 above)
Comment (rustyrussell) [reply channeld/channel_wire.csv]: That would have been more work, I just used sed on each daemon :)    But yeah, could just insert this in the generator, then assert that type & DAEMON_BITS is correct; I have another patch done which makes generator treat the bolt specs different from out CSV files, for 32-bit lengths, so it wouldn't be hard to add later.

I've added a FIXME :)

File: onchaind/onchain.c (hunk 5 above)
Comment (rustyrussell) [reply onchaind/onchain.c]: Had to look at the src to figure out what you were saying.  Yep, fixed, though I'm not sure it's actually much clearer.

**Pull request merged**